### PR TITLE
Adding LLVM obuscation tutorial

### PR DESCRIPTION
**Pull request by artas360**
Adding tutorial & Modifying an outdated comment in ObfuscateZero.cpp

### CODE REVIEW

File: doc/llvm_obfuscation_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,968 @@
+============================================================
+Introduction on turning good code into attrocities with LLVM
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: Turning Regular Code into Attrocities with LLVM

File: doc/llvm_obfuscation_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,968 @@
...
+============================================================
+
+Objective
+=========
+The objective of this post is to explain the basics of LLVM bytecode obfuscation.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: through an out-of-source build single pass.

File: doc/llvm_obfuscation_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,968 @@
...
+*'But why obfuscate the LLVM bytecode? Why not the source code? Or the binary?'*, you may ask.
+
+Well it's because LLVM is *super swag* right now and using it is cool.
+But, regarding the engineering aspects, it is because there are lots of front-ends converting different languages into the same LLVM bytecode (Clang/Clang++ for C/C++, Mono LLVM for C#, Pyston for ... Python and so on).
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: no space before ...

File: doc/llvm_obfuscation_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,968 @@
...
+Hence by working at the bytecode level we can obfuscate programs produced by many languages without even knowing them.
+Another good thing is that the obfuscation can be easily integrated with the compilation chain.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: s/the compilation chain/ existing compilation chain/

File: doc/llvm_obfuscation_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,968 @@
...
+        prime1 * ((x | any1)**2) != prime2 * ((y | any2)**2)
+
+    - prime1 and prime2 are *distinct* prime numbers
+    - any1 and any2 are *distinct* strictly positive random numbers
+    - x and y are two variables picked from the program (they have to be reachable from the obfuscation instructions)
+
+This expression will always return a boolean zero (false). The idea is to insert this test into our code, just before the 0 we want to obfuscate and to replace this 0 by the result of our comparison.
+As you have probably noticed we will have to pay attention to the type of the original 0 and make sure we cast the result of our expression to its type.
+
+This obfuscation may not be the most sophisticated ever written but it's enough to learn the basics of LLVM bytecode obfuscation (and to annoy our reverse engineer friends.)
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: and maybe to annoy our .. friends for a few minutes, until they issue a `miasm` script to clean this a bit :-)

File: doc/llvm_obfuscation_tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,968 @@
...
+    * And... wait....
+        ``>IRL sleep 1000``
+
+    * (Optional) You should set your path to include the freshly baked LLVM tools and clang:
+        ``>$ export PATH=~path/to/llvm/build/llvm/bin:$PATH``
+
+    * (Optional) You can run the test suite to make sure that LLVM was built correctly:
+        ``>$ make check``
+
+You have just built LLVM and clang but we going to build the passes out of the LLVM source tree. To do so we have prepared a git repository with the basic infrastructure.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: we going -> we are going

File: doc/llvm_obfuscation_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,968 @@
...
+
+        ``>$ git clone https://github.com/quarkslab/llvm-passes``
+
+From now on we will be working exclusively inside the ``llvm-passes`` folder (we will refer to it as ``$PASSDIR``). So let's visit our new office:
+    * *cmake*: cmake definitions to check the python environment. Required to generate our passes test suites.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: s/python/Python/

Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: to link to the LLVM environment and also check the Python environment for a few extra tools (we are using `lit (ref)` to run the validation. Wait. Did I wrote validation?)

File: doc/llvm_obfuscation_tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,968 @@
...
+                                   PassManagerBase &PM) {
+      PM.add(new MyPass());
+    }
+    static RegisterStandardPasses
+        RegisterMBAPass(PassManagerBuilder::EP_EarlyAsPossible,
+                        registerMyPassPass);
+
+If you have been paying attention so far you should remember that we are going to obfuscate null literals.
+And to do so we will randomly pick two variables reachable from where the replacement occurs.
+So, in order to keep the pass as simple as possible we are going to work at the basic bloc level, this way there will be no reachability problems with the variables we encounter.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst:152]: This could be greatly enhanced using dominators http://llvm.org/docs/doxygen/html/classllvm_1_1DominatorTree.html, but that's for another story!

File: doc/llvm_obfuscation_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,968 @@
...
+This is why our class derives from the ``BasicBlockPass`` class.
+
+.. code:: C++
+
+    class MyPass : public BasicBlockPass
+
+
+Do or do not there is no ... test
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: no space before...

File: doc/llvm_obfuscation_tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,968 @@
...
+    int foo(){return 1;}
+
+    int main() {
+        printf("Hello world\n");
+
+        return 0;
+    }
+
+
+And compile it with our awesome pass using:
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: .. code:: bash

```
  >$ opt -load $PASSDIR/build/llvm-passes/LLVMMyPass.so -my-pass in.ll -o out.bc
```

our directly with clang:

File: doc/llvm_obfuscation_tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,968 @@
...
+If any of the operand is a null literal we print a message on the debug stream and we continue.
+You may have noticed the for loop is initialized with ``BB.getFirstInsertionPt()``.
+We could have iterated through the block with a foreach like:
+
+.. code:: C++
+
+    for(auto &I : BB) {
+    }
+
+But we do not want to modify the special instructions located at the begining of the block (the `phi instructions <http://en.wikipedia.org/wiki/Static_single_assignment_form#Converting_out_of_SSA_form>`_), so we skip them altogether and set the iterator to the first 'normal' instruction.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: s/the/ some/ special instructions

File: doc/llvm_obfuscation_tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,968 @@
...
+
+The ``isValidCandidateOperand`` method checks if its parameter is a literal (constant means literal in LLVM, not variable declared ``const``).
+It also checks the type of the literal, it must not be a pointer or a floating point value (you will see later why).
+The type checks are done with the ``isa<>`` function which checks if its parameter can be cast to the type given by the template parameter.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: It's used in LLVM instead of RTTI(run time type information) because it was deemed to expensive http://llvm.org/docs/CodingStandards.html#do-not-use-rtti-or-exceptions

File: doc/llvm_obfuscation_tutorial.rst [hunk 13]
```diff
@@ -0,0 +1,968 @@
...
+    ; Function Attrs: nounwind uwtable
+    define i32 @main() #0 {
+      %1 = alloca i32, align 4
+      store i32 0, i32* %1
+      %2 = call i32 @puts(i8* getelementptr inbounds ([13 x i8]* @.str, i32 0, i32 0))
+      ret i32 0
+    }
+
+The two 0 that triggered the debug message from our pass are in the ``store`` and ``ret`` instructions.
+As you can see the compilation to bytecode introduced an expression not present in the source code. This will probably happen quite often when developing LLVM passes.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: s/ This is probaly.../ This is obviously due to the SSA (ref) form used by the IR.

File: doc/llvm_obfuscation_tutorial.rst [hunk 14]
```diff
@@ -0,0 +1,968 @@
...
+While debugging your future passes you will probably have to read a lot of bytecode so you should familiarize yourself with it.
+Lucky for you it's pretty easy to read (at least compared to asm).
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: BTW, you noticed the IR is strongly typed. this helps a lot!

File: doc/llvm_obfuscation_tutorial.rst [hunk 15]
```diff
@@ -0,0 +1,968 @@
...
+
+    1. To know the variables reachable from the instruction we replace
+    2. To generate the instructions of the arithmetic expression seen earlier
+    3. To insert those expressions back into the code
+    4. (Optional) Generate random prime numbers
+
+Reachable variables
+~~~~~~~~~~~~~~~~~~~
+
+To be sure to have a pool of **reachable** variable during our obfuscation, we are going to register all the Integers we come across while we iterate through the bloc instructions.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: s/Integers/variables with integral type/

File: doc/llvm_obfuscation_tutorial.rst [hunk 16]
```diff
@@ -0,0 +1,968 @@
...
+      %3 = load i32* %a, align 4
+      %4 = mul nsw i32 %3, 3
+      store i32 %4, i32* %a, align 4
+      ret i32 0
+    }
+
+
+There are a few things that you should remember from this little modification:
+* The LLVM bytecode is in `SSA form <http://en.wikipedia.org/wiki/Static_single_assignment_form>`_, so you will see variables that you didn't explicitly declared appear in the bytecode. Typically temporary result or ``load``s.
+* A variable declaration in your code returns a **pointer** in the bytecode not an instance of the type of the variable.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: Clang translates variable declaration into variable allocated on the stack (through the alloca instruction ). A later pass (Mem2reg) takes care of putting them in registers when possible.

File: doc/llvm_obfuscation_tutorial.rst [hunk 17]
```diff
@@ -0,0 +1,968 @@
...
+* You *need* to look at the bytecode to understand what you're *actually* telling LLVM to do (at least at first :p).
+* Using the ``operator<<`` is **very** useful for debug. (You can even use it on blocks, functions, ...)
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: The return value of `errs()` is overloaded for most LLVM types, so use it!

File: doc/llvm_obfuscation_tutorial.rst [hunk 18]
```diff
@@ -0,0 +1,968 @@
...
+
+.. code:: C++
+
+    prime_type p1 = getPrime(),
+               p2 = getPrime(p1);
+
+    if(p2 == 0 || p1 == 0)
+        return nullptr;
+
+The LLVM bytecode is strongly typed so will need to play a little with the types.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: we will need

File: doc/llvm_obfuscation_tutorial.rst [hunk 19]
```diff
@@ -0,0 +1,968 @@
...
+              errs() << "MyPass: could not rand pick a variable for replacement\n";
+            }
+          }
+        }
+      registerInteger(Inst);
+    }
+    return modified;
+  }
+
+and here is the code full code (with the random prime generation):
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: s/generation/tabulation/

File: doc/llvm_obfuscation_tutorial.rst [hunk 20]
```diff
@@ -0,0 +1,968 @@
...
+    }
+
+In LLVM bytecode access to structure members turns into the ``GetElementPointer`` instruction. It looks like this:
+
+.. code:: llvm
+
+    %4 = getelementptr inbounds %struct.s* %s1, i32 0, i32 0
+
+As you can there are to integer operands at the end. The first one is used when going through an array, so in our case it will always be 0.
+The second one is the index of structure member we are accessing. If you access ``s.a`` it will be 0 and it will be 1 for ``s.b``.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst:890]: for more info on getelementptr, see http://llvm.org/docs/GetElementPtr.html

File: doc/llvm_obfuscation_tutorial.rst [hunk 21]
```diff
@@ -0,0 +1,968 @@
...
+Pretty easy, no? Well the hard part is that this kind of problems is almost impossible to anticipate unless you know all the LLVM instructions.
+The only solution to find this id to run your passes on big projects, see where it crashes and find out why.
+
+Your code should now be the same as the ObfuscateZero pass.
+And since I don't want to dump all the code on this page (again) from now on we are going to use the ObfuscateZero pass for our tests.
+
+Tests, tests and more tests
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+I hope this last part made you understand that validation is crutial before using your obfuscations in prod.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst]: s/crutial/critical/

File: doc/llvm_obfuscation_tutorial.rst [hunk 22]
```diff
@@ -0,0 +1,968 @@
...
+For ObfuscateZero we used the ``llvm-lit`` testing tool, shipped with LLVM.
+This tool runs the tests you specify with a particular syntax. Take a look in the files in the ``$PASSDIR/tests/ObfuscateZero`` folder to learn how to use it.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst:960]: I'd rather talk about lit, and `pip install lit`

File: doc/llvm_obfuscation_tutorial.rst [hunk 23]
```diff
@@ -0,0 +1,968 @@
...
+This will run the ``ObfuscateZero`` tests, which you can modify to test your pass. But it's going to take some time.
+To validate ``ObfuscateZero`` we also compiled a C++ code since some constructs are not present when compiling from C.
+However the test file has not been shipped in the git.
+
+Bye Bye
+=======
+
+This tutorial was just an introduction to writing LLVM passes and using them for obfuscation.
+There are many more funny things to do to make your code very annoying for reverse engineers.
+I hope this will help you get started.
```
Comment (serge-sans-paille) [inline doc/llvm_obfuscation_tutorial.rst:981]: But remember, if you choose the quick and easy path as Vader did - you will become an agent of evil.

**Comment by artas360** [comment]
@serge-sans-paille fixed

**Comment by serge-sans-paille** [comment]
paradise yeah!

File: doc/llvm_obfuscation_tutorial.rst [hunk 24]
```diff
@@ -0,0 +1,979 @@
...
+
+    * Now download the clang sources inside the llvm_src/tools directory.
+        ``>$ cd llvm_src/tools && git clone --depth=1 --branch=release_35 https://github.com/llvm-mirror/clang``
+
+    * Create a build directory somewhere *out of the llvm source tree*:
+        ``>$ mkdir path/to/llvm/build/llvm``
+
+    * Let's build LLVM:
+        ``>$ cd ~/llvm && cmake path/to/llvm/sources``
+        ``>$ make -j``
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: Can you clean-up all the paths, it is hard to follow.

File: doc/llvm_obfuscation_tutorial.rst [hunk 25]
```diff
@@ -0,0 +1,979 @@
...
+    * And... wait....
+        ``>IRL sleep 1000``
+
+    * (Optional) You should set your path to include the freshly baked LLVM tools and clang:
+        ``>$ export PATH=~path/to/llvm/build/llvm/bin:$PATH``
+
+    * (Optional) You can run the test suite to make sure that LLVM was built correctly:
+        ``>$ make check``
+    * If you want to run the existing tests you first need to install `pip <https://pypi.python.org/pypi/pip>`_ (python-pip). Then use it to install lit:
+        ``>$ pip install lit``
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: I doubt it is needed?

File: doc/llvm_obfuscation_tutorial.rst (hunk 8 above)
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:152]: even better, generating different patterns in multiple basic blocks and using phi to select, or even part of the pattern in various block and recomposing it in various paths. It might increase the entropy?

File: doc/llvm_obfuscation_tutorial.rst [hunk 26]
```diff
@@ -0,0 +1,979 @@
...
+        return 0;
+    }
+
+
+
+And compile it with our awesome pass using:
+
+.. code:: bash
+
+    >$ clang++ -Xclang -load -Xclang $PASSDIR/build/llvm-passes/LLVMMyPass.so path/to/test/file -o /tmp/awesome.out
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: why not generating a .ll directly?

File: doc/llvm_obfuscation_tutorial.rst [hunk 27]
```diff
@@ -0,0 +1,979 @@
...
+
+Or if you already have a file with the LLVM bytecode:
+
+.. code:: bash
+
+  >$ opt -load $PASSDIR/build/llvm-passes/LLVMMyPass.so -MyPass in.ll -o out.bc
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: -S allows to have a .ll and readable output. No need for bytecode at this point.

File: doc/llvm_obfuscation_tutorial.rst [hunk 28]
```diff
@@ -0,0 +1,979 @@
...
+To find the null literals we need to iterate over every instruction of the block and check if one of the operands is null.
+
+.. code:: C++
+
+  //Add the following to your headers
+  #include "llvm/IR/Constants.h"
+  #include "llvm/IR/Instructions.h"
+
+  //Add the following to MyPass
+  virtual bool runOnBasicBlock(BasicBlock &BB) override {
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: Remove virtual here.

File: doc/llvm_obfuscation_tutorial.rst [hunk 29]
```diff
@@ -0,0 +1,979 @@
...
+        // dbgs() << "Ignoring Floats 0\n";
+        return nullptr;
+      } else if (C->isNullValue()) {
+        return C;
+      } else {
+        return nullptr;
+      }
+    } else {
+      return nullptr;
+    }
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:270]:
```
Constant *C;
if (!C = dyn_cast<Constant>(V)) return nullptr;
...
```

File: doc/llvm_obfuscation_tutorial.rst [hunk 30]
```diff
@@ -0,0 +1,979 @@
...
+    // parameters
+    for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),
+                                       end = BB.end();
+         I != end; ++I) {
+      Instruction &Inst = *I;
+      // We are not using an iterator because we will need i later.
+      for (size_t i{0}; i < Inst.getNumOperands(); ++i) {
+        if (Constant *C = isValidCandidateOperand(Inst.getOperand(i))) {
+          errs() << "I've found one sir!\n";
+        }
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:256]: This loop is ugly, `size_t i{0}` really?

```
for (Operand &Op : Inst.operands()) {
  if (Constant *C = isValidCandidateOperand(Op)) {
    ....
  }
}
```

File: doc/llvm_obfuscation_tutorial.rst [hunk 31]
```diff
@@ -0,0 +1,979 @@
...
+      }
+    }
+    return false;
+   }
+
+  Constant *isValidCandidateOperand(Value *V) {
+    if (Constant *C = dyn_cast<Constant>(V)) {
+      // Checking constant eligibility
+      if (isa<PointerType>(C->getType())) {
+        // dbgs() << "Ignoring NULL pointers\n";
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: Why NULL? Just pointer here right?

Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: Oh I got it, these are "sanity checks".
I'd rather write:

```
if (!C->isNullValue() || 
    C->getType()->isFloatingPointTy() ||
    C->getType()->isPointerTy())
    return nullptr;
// with the other change below, all the check are done earlier and here 
// you are at the minimum indentation level. Easier control flow.

  ....

```

Note the use of `C->getType()->isPointerTy()` for consistency with the floating point check.

File: doc/llvm_obfuscation_tutorial.rst [hunk 32]
```diff
@@ -0,0 +1,979 @@
...
+We could have iterated through the block with a foreach like:
+
+.. code:: C++
+
+    for(auto &I : BB) {
+    }
+
+But we do not want to modify some of the special instructions located at the begining of the block (the `phi instructions <http://en.wikipedia.org/wiki/Static_single_assignment_form#Converting_out_of_SSA_form>`_), so we skip them altogether and set the iterator to the first 'normal' instruction.
+
+The ``isValidCandidateOperand`` method checks if its parameter is a literal (constant means literal in LLVM, not variable declared ``const``).
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:286]: I think here it checks for NULL only, not any literal.

File: doc/llvm_obfuscation_tutorial.rst [hunk 33]
```diff
@@ -0,0 +1,979 @@
...
+It also checks the type of the literal, it must not be a pointer or a floating point value (you will see later why).
+The type checks are done with the ``isa<>`` function which checks if its parameter can be cast to the type given by the template parameter.
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: `isa` being not nice here, replace with the reference to the `dyn_cast``

File: doc/llvm_obfuscation_tutorial.rst [hunk 34]
```diff
@@ -0,0 +1,979 @@
...
+
+    ; Function Attrs: nounwind uwtable
+    define i32 @foo() #0 {
+      ret i32 1
+    }
+
+    ; Function Attrs: nounwind uwtable
+    define i32 @main() #0 {
+      %1 = alloca i32, align 4 ; This instruction...
+      store i32 0, i32* %1     ; ... and this one are useless, they would be deleted if we used an optimization flag.
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:313]: They are not useless taken separately, they produce the identity and can be thrown out when they come together (if the other uses are not too complex)

File: doc/llvm_obfuscation_tutorial.rst [hunk 35]
```diff
@@ -0,0 +1,979 @@
...
+        // parameters
+        for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),
+                                           end = BB.end();
+             I != end; ++I) {
+          Instruction &Inst = *I;
+            for (size_t i{0}; i < Inst.getNumOperands(); ++i) {
+              if (Constant *C = isValidCandidateOperand(Inst.getOperand(i))) {
+                errs() << "I've found one sir!\n";
+              }
+          }
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: indentation is messy

File: doc/llvm_obfuscation_tutorial.rst [hunk 36]
```diff
@@ -0,0 +1,979 @@
...
+          registerInteger(Inst);
+        }
+        return false;
+      }
+
+    private:
+      void registerInteger(Value &V) {
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:382]: LLVM coding standard ask for a doxygen description :)

File: doc/llvm_obfuscation_tutorial.rst [hunk 37]
```diff
@@ -0,0 +1,979 @@
...
+        if (V.getType()->isIntegerTy()) {
+          IntegerVect.emplace_back(&V);
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: emplace_back a pointer does not make a difference with push_back, except that the reader will double check the container to be sure because emplace_back is unusual.

File: doc/llvm_obfuscation_tutorial.rst [hunk 38]
```diff
@@ -0,0 +1,979 @@
...
+      store i32 %4, i32* %a, align 4
+      ret i32 0
+    }
+
+
+There are a few things that you should remember from this little modification:
+    * The LLVM bytecode is in `SSA form <http://en.wikipedia.org/wiki/Static_single_assignment_form>`_, so you will see variables that you didn't explicitly declared appear in the bytecode. Typically temporary result or ``loads``.
+    * A variable declaration in your code returns a **pointer** in the bytecode not an instance of the type of the variable. This is because Clang translates variable declarations into variables allocated on the stack (through the ``alloca`` instruction). A later pass (Mem2reg) takes care of putting them in registers when possible.
+    * You *need* to look at the bytecode to understand what you're *actually* telling LLVM to do (at least at first :p).
+    * The return value of errs() is overloaded for most LLVM types, so use it! This is **very** useful for debug. (You can even use it on blocks, functions, ...)
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: `dbgs()` is better than `errs()` cause easier to turn on/off per pass.

File: doc/llvm_obfuscation_tutorial.rst [hunk 39]
```diff
@@ -0,0 +1,979 @@
...
+Well it's because LLVM is *super swag* right now and using it is cool.
+But, regarding the engineering aspects, it is because there are lots of front-ends converting different languages into the same LLVM bytecode (Clang/Clang++ for C/C++, Mono LLVM for C#, Pyston for... Python and so on).
+Hence by working at the bytecode level we can obfuscate programs produced by many languages without even knowing them.
+Another good thing is that the obfuscation can be easily integrated with the existing compilation chain.
+
+Now let's talk about what we're going to do.
+Our mission (and we have no choice but to accept it) is to obfuscate all null literals in the code.
+It means that we are going to replace (almost) all the zeroes in the code by a non-trivial boolean expression always false.
+
+        prime1 * ((x | any1)**2) != prime2 * ((y | any2)**2)
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:20]: What about overflow and undefined behavior?

File: doc/llvm_obfuscation_tutorial.rst [hunk 40]
```diff
@@ -0,0 +1,979 @@
...
+.. code:: C++
+
+    // Creating the LLVM objects representing literals
+    Constant *any1 = ConstantInt::get(IntermediaryType, 1 + RandAny(Generator)),
+             *any2 = ConstantInt::get(IntermediaryType, 1 + RandAny(Generator)),
+             *prime1 = ConstantInt::get(IntermediaryType, p1),
+             *prime2 = ConstantInt::get(IntermediaryType, p2),
+             // Bitmasks to prevent overflow
+             *OverflowMaskLHS = ConstantInt::get(IntermediaryType, 0x00000007),
+             *OverflowMaskRHS = ConstantInt::get(IntermediaryType, 0x00000007);
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: Since the mask is the same, creating it only once should be OK?

File: doc/llvm_obfuscation_tutorial.rst (hunk 8 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst:152]: This is an intriduction so it's probably better to KISS, no?

File: doc/llvm_obfuscation_tutorial.rst (hunk 24 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: My bad, I'll do that

File: doc/llvm_obfuscation_tutorial.rst (hunk 25 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: Easier for automatic testing with cmake (that's why I understood at least...)

File: doc/llvm_obfuscation_tutorial.rst (hunk 26 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: To let them execute the code once, at least they know how to do it.

File: doc/llvm_obfuscation_tutorial.rst (hunk 27 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: I ll let serge comment on this one

File: doc/llvm_obfuscation_tutorial.rst (hunk 30 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst:256]: But it looks cool :'(

Nop, I need the index to reinject the new operand in the inst.

File: doc/llvm_obfuscation_tutorial.rst (hunk 31 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: I do agree that it's ugly but this way you can have different debug messages for each inst. 

If you realy think that's not usefull I'll change it but at least you see what's happening during debug.

File: doc/llvm_obfuscation_tutorial.rst (hunk 32 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst:286]: the first check is `if (Constant *C = dyn_cast<Constant>(V))`, unless I misunderstood sthg that checks for literals, no?

File: doc/llvm_obfuscation_tutorial.rst (hunk 33 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: I'm not sure what you mean. Could you explain a bit please?

File: doc/llvm_obfuscation_tutorial.rst (hunk 35 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: That's the clang formating tool, serge asked me to rely on it.

File: doc/llvm_obfuscation_tutorial.rst (hunk 37 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: Yeah I know but I like emplace back :D.
Maybe I ll change it...

File: doc/llvm_obfuscation_tutorial.rst (hunk 39 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst:20]: Later in the tuto

File: doc/llvm_obfuscation_tutorial.rst (hunk 40 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: Yep, it was just for symmetry

File: doc/llvm_obfuscation_tutorial.rst (hunk 30 above)
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:256]: There no reuse in this version of the code ;)

but seriously forget the curly braces initialization for the loop index...

File: doc/llvm_obfuscation_tutorial.rst (hunk 31 above)
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: Right now your debug is comment out.

Since I read function from the begining to the end and not the opposite usually, I'd rather do:

```
Constant *C;
if (!C = dyn_cast<Constant>(V)) return nullptr;
if (!C->isNullValue()) return nullptr;
// We found a NULL constant, lets validate it

if (C->getType()->isFloatingPointTy()) {
  dbgs() << "Ignoring Floats 0.0\n";
  return nullptr;
}
if(C->getType()->isPointerTy()) {
  dbgs() << "Ignoring NULL pointer\n";
  return nullptr;
}
....
return C;
```

Note also, if you just want an integer, would this be enough?

```
Constant *C;
if (!C = dyn_cast<Constant>(V)) return nullptr;
if (!C->isNullValue()) return nullptr;
// We found a NULL constant, lets validate it
if(!C->getType()->isIntegerTy()) {
  dbgs() << "Ignoring non integer value\n";
  return nullptr;
}
return C;
```

File: doc/llvm_obfuscation_tutorial.rst (hunk 32 above)
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:286]: OK, I took the sentence alone, while the following sentences are also part of the description.

File: doc/llvm_obfuscation_tutorial.rst (hunk 35 above)
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: I doubt clang-format produces:

```
            for (size_t i{0}; i < Inst.getNumOperands(); ++i) {
              if (Constant *C = isValidCandidateOperand(Inst.getOperand(i))) {
                errs() << "I've found one sir!\n";
              }
          }
```

There has been copy/paste involved somehow.

File: doc/llvm_obfuscation_tutorial.rst [hunk 41]
```diff
@@ -0,0 +1,979 @@
...
+
+      void registerInteger(Value &V) {
+        if (V.getType()->isIntegerTy())
+          IntegerVect.emplace_back(&V);
+      }
+
+      // Return a random prime number not equal to DifferentFrom
+      // If an error occurs returns 0
+      prime_type getPrime(prime_type DifferentFrom = 0) {
+          static std::uniform_int_distribution<prime_type> Rand(0, sizeof(Prime_array) / sizeof(prime_type));
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: I doubt this went through clang-format? (80 cols)

File: doc/llvm_obfuscation_tutorial.rst [hunk 42]
```diff
@@ -0,0 +1,979 @@
...
+
+In this case the compiler has optimized out ``a`` which was the only integer available for the obfuscation.
+This explains why the obfuscation aborted.
+
+Even if it is frustrating it is not a real problem, since the compiler won't delete all the potential integer in a real code.
+However this is very annoying when writing tests.
+The easiest work-around is to declare ``volatile`` the variables you don't want to be optimized out.
+
+You might think that not using the optimizer is a good solution but:
+    * If your obfuscation can't resist an optimizer, it won't resist reverse engineers.
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst:855]: In this case it does not seem that the obsfucation han't resisted the optimizer but that the optimizer ran before the obfuscation.

File: doc/llvm_obfuscation_tutorial.rst [hunk 43]
```diff
@@ -0,0 +1,979 @@
...
+        return 0;
+    }
+
+In LLVM bytecode access to structure members turns into the ``GetElementPointer`` instruction. It looks like this:
+
+.. code:: llvm
+
+    %4 = getelementptr inbounds %struct.s* %s1, i32 0, i32 0
+
+As you can there are to integer operands at the end. The first one is used when going through an array, so in our case it will always be 0.
```
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: As you can ....
There are _to_

File: doc/llvm_obfuscation_tutorial.rst (hunk 33 above)
Comment (joker-eph) [inline doc/llvm_obfuscation_tutorial.rst]: Because I said earlier to replace the `isa<>` for the pointer to `C->getType()->isPointerTy()`. This sentence has to be changed to  refer to dyn_cast to talk about LLVM-Style RTTI

File: doc/llvm_obfuscation_tutorial.rst (hunk 42 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst:855]: Yeah this is explained earlier, it's more like general remark.
If you think I should clarify let me know.

File: doc/llvm_obfuscation_tutorial.rst (hunk 43 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: Anyway to make all traces of this disappear :)?

File: doc/llvm_obfuscation_tutorial.rst (hunk 35 above)
Comment (artas360) [inline doc/llvm_obfuscation_tutorial.rst]: Oups, sorry. I thought you were speaking about : 

```
for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),
end = BB.end();
I != end; ++I) {
```
